esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,470 @@
+//! # I2S in Camera Slave receiving mode.
+//!
+//! ## Overview
+//! The I2S peripheral supports a camera slave mode for high-speed data
+//! transfer from external camera modules.
+//! The driver mandates DMA for efficient data transfer.
+//!
+//! ## Examples
+//! ```rust, no_run
+#![doc = crate::before_snippet!()]
+//! # use esp_hal::i2s::master::camera::{Camera, Config};
+//! # use esp_hal::dma_rx_stream_buffer;
+//!
+//! # const BUF_SIZE: usize = 20 * 1000;
+//! # let dma_rx_buf = dma_rx_stream_buffer!(BUF_SIZE, 1000);
+//!
+//! let cam_siod = peripherals.GPIO26;
+//! let cam_sioc = peripherals.GPIO27;
+//! let cam_xclk = peripherals.GPIO32;
+//!
+//! let camera = Camera::new(
+//!     peripherals.I2S0,
+//!     Config::default(),
+//!     peripherals.DMA_I2S0,
+//! )
+//! .with_ws(peripherals.GPIO22)
+//! .with_vsync(peripherals.GPIO25)
+//! .with_hsync(peripherals.GPIO23)
+//! .with_data_pins(
+//!     peripherals.GPIO5,
+//!     peripherals.GPIO18,
+//!     peripherals.GPIO19,
+//!     peripherals.GPIO21,
+//!     peripherals.GPIO36,
+//!     peripherals.GPIO39,
+//!     peripherals.GPIO34,
+//!     peripherals.GPIO35,
+//! );
+//!
+//! let mut transfer = camera.receive(dma_rx_buf, BUF_SIZE).unwrap();
+//!
+//! # Ok(())
+//! # }
+//! ```
+
+use core::{
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    dma::{ChannelRx, DmaEligible, DmaError, DmaRxBuffer, PeripheralRxChannel, Rx, RxChannelFor},
+    gpio::{InputPin, Level, Pull},
+    i2s::master::{Error, ExtendedSignals, RegisterAccess},
+    peripheral::{Peripheral, PeripheralRef},
+    system::PeripheralClockControl,
+    Blocking,
+};
+
+/// Supported data formats
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum DataFormat {
+    /// camera sends byte sequence: s1, s2, s3, s4, ...
+    /// fifo receives: 00 s1 00 s2, 00 s2 00 s3, 00 s3 00 s4, ...
+    DualChannel16   = 0,
+    /// camera sends byte sequence: s1, s2, s3, s4, ...
+    /// fifo receives: 00 s1 00 s2, 00 s3 00 s4, ...
+    SingleChannel16 = 1,
+    // DualChannel32   = 2, // not present in https://github.com/espressif/esp32-camera/blob/4467667b71f22a4c7db226f24105a9350abe7a05/target/esp32/ll_cam.c#L66
+    /// camera sends byte sequence: s1, s2, s3, s4, ...
+    /// fifo receives: 00 s1 00 00, 00 s2 00 00, 00 s3 00 00, ...
+    SingleChannel32 = 3,
+}
+
+/// Represents the camera interface.
+pub struct Camera<'d, I2S: DmaEligible> {
```

Comment:
The `I2S` generic parameter should be erased, like this https://github.com/esp-rs/esp-hal/blob/7c159b6070b9025903de7d24ac600839625a191f/esp-hal/src/i2s/parallel.rs#L237-L244

Also, do the guard thing.